### ISSUE TITLE
Provide minimal pid1 application or Split pid1 tasks and service management into 2 applications

### ISSUE THREAD

**Issue by oiaohm** [systemd/systemd#6323]
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with
All-233
> …

### Used distribution

> …

### In case of bug report: Expected behaviour you didn't see

> …
2 Binaries one being init and 1 being service management.
### In case of bug report: Unexpected behaviour you saw
Excess code in containers.  Existing issues with replacing init with bash and the like.

https://www.fpcomplete.com/blog/2016/10/docker-demons-pid1-orphans-zombies-signals

The issue is covered above with docker.   That you need the zombie reaping functionality.   Or bad things can happen.

Now if the PID 1 process is stripped to bone and all it does is start something be it systemd service management or some service starting in a container or  bash when thing have gone wrong things are nicer.   This is not a feature sysvinit had.   Yes init=value on linux command line is useful to a point but using this with bash is not clean to do a repair and going to operational system.

Minimal pid1 executable is required for containers that are really only running 1 service.   This is attack surface reduction.

PID1 application and service management SMF in solaris for basically the same reasons is two binaries the PID1 application is reused inside solaris zones.

Like it or not this is a real design flaw.  That has been in history ignored.   This is not a design flaw unique to systemd.   sysvinit is guilty of providing not providing way to cleanly start a task for issue to be repaired and then continue the system this was historically hacked around by run levels.   Lot of init systems are guilty of this problem.

Zombie reaper that can start selected application from the kernel command line should have really always existed.

Also the design of the Zombie reaper should be very tight.   So Zombie reaper should be a static binary that is independent of libc that can be audited massively in ideal case.    This is kind of important like how do you get into system if a update to libc or kernel has broken something that systemd current PID1 executable uses.

If systemd is being designed to cover all boot and container cases a minimal PID1 should exist in the source base.

Using too big of binary used for PID1 interferes with system recovery processes by providing large of a surface area to cause failure as systemd currently does.   Sysvinit depending on shell to run tasks to get to a recovery runlevel was also has the issue large of surface area to cause failure just it was not all in PID1 task binary so kind of hidden from view.   

init=/bin/bash is a case of not enough code so risks bleeding zombies while you are attempting to repair system.   So all 3 existing options are nicely broken.

There is a big issue around systemd of sealing off bugs that bring up real inconvenient issues.    It does not take long to sit down and start calculating recovery path and volume of code dependency.   The more lines of code a recovery path depends on the more likely it will have a defect and not be workable recovery mode when you need it.   There is also the too minimal option that you have used less code but there is now not enough code to provide properly function.    

This was and has not done in the systemd or sysvinit or most linux service management system design processes at this stage.   If this had been done for systemd a application for PID1 that is very compact would exist in the systemd code base.

I am not saying that /lib/systemd/systemd as  PID1 is wrong for a system that is functional the reality is it not usable when systems are broken.   Reality of this world is stuff breaks.    If a min PID1 existed it could be provided as option for those who don't like dynamic binaries in PID1 so reducing hate against project. 

This is a fault systemd has copied from sysvinit and the like in a more visible way.   When you are attempting to be better is foolish to be coping mistakes.   The complains that broken systemd is hard to recover is true this existing mistake makes it harder than it should be.  Reason for splitting service management and pid1 is to minimise the amount of change person doing recovery needs to-do.   Of course this something the project may need to debate how recovery and system operation should appear.   Recovery model needs work.

**Comment by floppym** [comment]
Other people have already written minimal init daemons for use in containers (tini, dumb-init, etc). Why should systemd reinvent the wheel?

**Comment by oiaohm** [comment]
floppym its recovery.   You really do need a minimal init to start bash so you can do a repair in bash and then continue the system.   Now if you are needing to work around a hardware fault that is a little unstable rebooting may not be valid option.

Also if systemd is claim is to be a complete system lacking a minimal init is a defect.   Really tini, dumb-init etc are all reinventing wheel because the core init system that distributions provide are not coming with this already.    Thinking its a tool for recovery as well as container init the item does two jobs.  

It can be doing 3 jobs.   The third be minimising volume of code running with PID1 privilege process number.

Basically you have the reinventing wheel question backwards.  Why are tini, dumb-int etc needing to reinvent the wheel in the first place when what they are doing should be a feature of a properly designed init system,   The answer is init systems for a long time have not been properly designed.

**Comment by poettering** [comment]
I am sorry, but I see no value in splitting this up. PID 1 has special semantics on UNIX, and we need these special semantics to do proper service management (as PID 1 is the process getting SIGCHLD for all processes that have no explicit parent). Any split of this functionality would be entirely artificial and require us to come up with a complex IPC logic that would relay the SIGCHLD events from PID 1 to PID 2 and needless IPC adds complexity and fragility, and doesn't help making things simpler. In fact, such a split would result in *more* code and in *more* complexity than the current simple approach.

Moreover, the process reaping thing is a weakness of Docker, and should be fixed there, not worked around through systemd, that's really not our business to be in. It's pretty easy to fix this in Docker, for example by implementing the equivalent to systemd-nspawn's `--as-pid2` switch in it: the PID 1 stub should really sit in Docker, and not outside of it.

Then, even if Docker doesn't fix this (I don't think they even grok the problem): it's trivial to work-around this outside of the systemd project, and to my knowledge there are multiple other projects which already solve this, hence why stick that into systemd a second time?

Also, systemd as PID 1 is tiny, really tiny with a relatively small attack surface, and very little privileges if you compare it with the piece of software that runs directly beneath it: the Linux kernel: it's huge, it's way more complex, runs at much higher privileges, and has a much higher attack surface. I am not sure what the benefit of shifting around complexity in userspace is supposed to be, if it all falls apart because the kernel is *so* *much* *worse*. Or to say this differently, please work with the kernel folks to get the kernel split into pieces each running with different privileges, before trying to do this with PID1...

Also, note that service management is required to keep the system up, and hence, if that's done in PID2 and PID2 crashes for some reason, then the system is as fucked as it would be if this was done in PID1, hence nothing is gained. It's like the kernel: if the kernel crashes you are fucked. And much like the kernel is required to keep the system running, service management is.

Not that "emergency" mode in systemd is perfect for just getting a shell, and then continuing from there. If you boot with "emergency" on the kernel cmdline you'll get just PID 1 plus the shell and nothing else, but can then boot up the rest of the system from there. (That said, you don't even need that. You can also do init=/bin/sh and as last step do exec /usr/lib/systemd to continue booting).

Hence, sorry, but I very much disagree that this would help anyone. Docker isn't waiting for this, and we are not the place to fix Docker's shortcomings really. I really don't appreciate the added complexity for the notification relay and so on, as I am very interested in a simpler, more robust one, instead of a complex, fragile one.

Sorry.

**Comment by oiaohm** [comment]
_Also, note that service management is required to keep the system up, and hence, if that's done in PID2 and PID2 crashes for some reason, then the system is as fucked as it would be if this was done in PID1,_ 
This is you failing to think.   It depends if PID2 is restart-able/fixable by triggering another program.   So PID1 could be a watch dog on PID2.  So PID2 fails to give PID1 its regular notice it results in PID1 starting off the repair program to the service management.

So your statement here is lack of thinking of what PID1 could doing for the service management.   Thinking the service management is processing files and could have got jammed to like disc io error where a hard drive has stalled and is not responding.   Now without a watch dog sitting on top of the service management you are stuffed.   With a watchdog sitting on top IO errors like this happen and system can recover and log what went wrong.

Of course this idea of having a PID1 that is a watchdog is from systems designed for life and death.   People you have only seen the crap that has been on Linux have no clue about using PID1 as a watchdog for the service manager.   So there are reasons in design to split the PID1 and service manager so the system does not have to stop dead because service manager fails to process something.

_Any split of this functionality would be entirely artificial and require us to come up with a complex IPC logic that would relay the SIGCHLD events from PID 1 to PID 2 and needless IPC adds complexity and fragility, and doesn't help making things simpler._ 

This raises question why do you need this information from PID1 by IPC I don't see this as required but as a incompetent excuse not to implement stuff properly.  This information should be exposed in monitoring of the cgroups and namespace contents or does this still need more work in the Linux kernel.

https://lwn.net/Articles/697462/
The reality is when the service manager starts it moves PID1 into a special cgroup with a ebpf connected.  Done no special IPC between PID1 and service manager instead standard debug logging by ebpf and cgroups.   Service manager able to monitor closer exploiting what cgroups can do these days.   Also SIGCHLD is one of the reasons to split the service manager and PID1 to get more clear information what is going on in the system.   

 **SIGCHLD is sent to the parent process.**
Except where it lost parent and it sent to PID1.

So service manager should be parent to a lot of different processes.    So should be receiving SIGCHLD directly as part of normal operation.   So in split those sending SIGCHLD to PID1are processes that have lost parent and SIGCHLD coming into the process id of the service manager are from those that it has started .

So currently it would not be simple to see what SIGCHLD messages are normal process exist under the service manager and what are PID1 job of being a reaper.  Because by merging the two parts into 1 you have mixed the signal sources so you no longer have clear logging of what is going on without having to perform extra processing.

_Not that "emergency" mode in systemd is perfect for just getting a shell, and then continuing from there. If you boot with "emergency" on the kernel cmdline you'll get just PID 1 plus the shell and nothing else, but can then boot up the rest of the system from there._ 

True but this can fall in a complete heap when it turns out that a library is missing due to a failed update.  init=/bin/sh does not reap zombies properly.   Docker demos what is wrong with depending on running a normal program as PID1 when trying to recover a system you can be attempt to run programs that don't run properly so breeding zombies at a huge rate.   You need a init program as PID1.  When things go wrong you want the most minimal init you can have.

Emergency mode init requirement are simple.
1) Static linked.   No dynamic libraries
2) bare minimal kernel feature requirements
3) be able to be directed to start what ever program happens to still work on the system.   Systemd emergency mode is that weak that at times you have to use /bin/sh and run the zombie gauntlet

Providing a minimal init that can be used in emergency when everything has gone hell.
